lib/ansible/galaxy/collection/__init__.py

```diff
@@ -681,7 +674,7 @@ def install_collections(
     unsatisfied_requirements = set(
         chain.from_iterable(
             (
-                Requirement.from_dir_path(sub_coll, artifacts_manager)
+                Requirement.from_dir_path(to_bytes(sub_coll), artifacts_manager)
```

Comment:
This line is the real fix, since `from_dir_path` is meant to only accept bytes.  All other changes are just related to actually adding the annotations to `from_dir_path`

Reply:
Why bytes and not str, though? It's a bad API, I was going for proper strings where possible IIRC.

Reply:
probably because it is a 'path' and those are fed to functions/system calls that require bytes.